src/adapter/src/client.rs

```diff
@@ -811,30 +811,37 @@ impl SessionClient {
             cmd
         });
 
+        // SPOOKY ZONE! See the doc comment on `SessionTerminateGuard` for why this exists.
+        //
+        // tl;dr this guard prevents a race where a `Session` is returned on `rx` but never placed
+        // back in `self` because the Future returned by this function is concurrently dropped.
```

Comment:
can this doc explain where the `send_with_cancel` future is dropped? Do we abort tasks with closed connections, or does hyper?